### PR TITLE
fix: reduce NewsDO hot query scans

### PR DESCRIPTION
**Pull request by whoabuddy**
## Summary

Reduces the highest-dollar `agent-news` NewsDO row-read path from the April Cloudflare bill audit.

Changes:
- rewrites `/signals` to build dynamic SQL predicates instead of `(? IS NULL OR ...)` clauses
- removes the unbounded list `COUNT(*)`; returns bounded pagination metadata with `hasMore` while preserving numeric `total`
- defers `signal_tags` loading until after the page of signal IDs is selected
- rewrites `/signals/counts` and `/init` count queries to small per-status indexed queries instead of `OR` / `COALESCE` grouped scans
- adds migration 27 with hot-path composite indexes
- adds `docs/cloudflare-cost-runbook.md` with the rows-read verification plan

## Expected Cloudflare metric movement

Target metric: Durable Objects SQLite `rows_read` for the `agent-news` script / NewsDO namespace.

Baseline from `cloudflare-bill-audit-2026-04.md`:
- NewsDO rows read in April: 427.8B
- `agent-news` DO invocations in April: 5.1M
- average: about 84,000 rows read per invocation

Expected after deploy:
- hot listing/count invocations should drop to roughly 50-500 rows read per request class
- 24h and 48h production windows should show orders-of-magnitude lower rows_read / invocation

Verification window:
- before deploy: capture prior 24h rows_read and DO invocations
- 15-30 min after deploy: check `/api/signals`, `/api/signals/counts`, `/api/init` health and WARN/ERROR logs
- 24h and 48h after deploy: recapture rows_read / invocation and record it in the audit/runbook thread

## Tests

Passed:
- `npm run typecheck`
- `npm test -- src/__tests__/signals.test.ts src/__tests__/do-client.test.ts src/__tests__/signal-counts-since.test.ts src/__tests__/schema-migration.test.ts src/__tests__/home-page.test.ts`

Not clean / caveats:
- `npm run lint` currently reports pre-existing style warnings in unrelated files and existing warnings in `news-do.ts`.
- Full `npm test` hit repeated Cloudflare `workerd` RPC-handler internal errors and 30s timeouts in unrelated payment/classified tests; the focused signal/init suite above passed.

### CODE REVIEW

File: src/objects/news-do.ts [hunk 1]
```diff
@@ -2106,63 +2303,42 @@ export class NewsDO extends DurableObject<Env> {
...
-          status,
-          dateStart,
-          dateEnd
-        )
-        .toArray();
-      const total = (totalRows[0] as { total: number } | undefined)?.total ?? signals.length;
+      // Avoid the unbounded COUNT(*) that made every paginated list scan the
+      // whole table. `total` remains numeric for old callers, but is now a
+      // bounded lower bound that reveals only whether another page exists.
+      const total = offset + signals.length + (hasMore ? 1 : 0);
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply src/objects/news-do.ts]: Fixed in `677dc87`: empty out-of-range pages now return `total: 0` instead of echoing the requested offset, while non-empty pages still return the bounded lower-bound (`offset + returned + next-page sentinel`). Added integration coverage for first page, final page, and beyond-end offset behavior.

File: src/lib/types.ts [hunk 2]
```diff
@@ -420,6 +420,8 @@ export interface DOResult<T> {
   approval_cap?: ApprovalCapInfo;
   /** Present on paginated list responses — count of matching rows ignoring limit/offset */
   total?: number;
+  /** Present on bounded paginated list responses when one more page is known to exist */
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply src/lib/types.ts]: Fixed in `677dc87`: updated `DOResult.total` docs to state that hot paginated paths may return a bounded lower-bound rather than an exact count, and callers should use `hasMore` for next-page availability.

File: src/routes/signals.ts [hunk 3]
```diff
@@ -136,9 +136,10 @@ signalsRouter.get("/api/signals", signalReadRateLimit, async (c) => {
   c.header("X-Timezone", "UTC");
   const response = c.json({
     signals: transformed,
-    // Count of all rows matching the filter set, across all pages.
-    // The signals list page renders "Page X of N" off this number.
+    // Bounded lower-bound count. Avoids making every list request run an
+    // unbounded COUNT(*) in NewsDO while preserving the legacy numeric field.
     total,
+    hasMore,
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply src/routes/signals.ts:142]: Fixed in `677dc87`: added `/api/signals` integration coverage for `hasMore` and bounded `total` semantics across first page, final page, and out-of-range offset cases.

**Review by arc0btc — APPROVED** [review]
Rewrites the three hottest NewsDO query paths to eliminate the unbounded `COUNT(*)` and `GROUP BY` + tag JOIN on every list request, replacing them with dynamic predicates and a `limit+1` probe. Production-scale change (5.1M invocations/month, 84K rows/invocation baseline) — solid work.

**What works well:**
- The `limit+1` sentinel swap for pagination is the right call — avoids a full-table count on every request while preserving the `hasMore` signal that clients actually need
- `buildSignalListWhere` / `appendSignalScopeFilters` cleanly eliminate the `(? IS NULL OR col = ?)` pattern that blocked index usage
- Deferred `fetchSignalTags` after page selection is correct — we now only fetch tags for the N returned rows, not for every candidate row
- Composite indexes in migration 27 are well-matched to the new query shapes — `(status, created_at)` and `(status, reviewed_at, created_at)` directly serve the per-status count loops
- Tests cover the bounded-pagination semantics end-to-end including the beyond-last-page edge case
- The runbook in `docs/cloudflare-cost-runbook.md` with explicit before/after windows and rollback signals is exactly what a change like this needs

**[question] `total` semantics change — client compatibility** (`src/routes/signals.ts`)
The new `total` is a lower-bound approximation: `offset + signals.length + (hasMore ? 1 : 0)`. On a mid-stream page with 25 results at limit=10, page 1 returns `total=11`, page 2 returns `total=21`, page 3 returns `total=25`. Any frontend code rendering "Showing X of Y signals" or "Page N of M" will silently show inaccurate totals until the last page. The cost tradeoff is absolutely worth it — just worth confirming no existing client renders page-count UI off this field before deploying.

**[suggestion] `querySignalCountRows` already sums reviewed+fallback internally** (`src/objects/news-do.ts`, around line 2454)
The function pushes exactly one row per status (with reviewed+fallback already added), so the outer `counts[r.status] += ...` accumulates correctly but would behave identically as `=`. It reads slightly confusingly given that `queryBeatStatusCountRowsSince` (which genuinely returns multiple rows per beat/status) needs `+=` for real. A comment like `// each status appears once; += is safe but equivalent to =` would save a future reader from having to verify the return structure.

**[nit]** The five new indexes appear in both the inline `SCHEMA_SQL` block and `MIGRATION_SIGNAL_HOT_PATH_INDEXES_SQL`. The `IF NOT EXISTS` guards make this safe, and the dual-definition pattern (schema for fresh instances, migration for existing) is correct — just flagging it's intentional duplication, not a mistake.

**Operational note:** Arc's sensors call `/api/signals` and `/api/signals/counts` via `do-client.ts` on every dispatch cycle. The `listSignalsPage` return type now includes `hasMore` — our code will pick that up automatically. The `total` field semantics change doesn't affect us since we iterate pages by checking for empty results rather than computing page counts from totals.
